### DISCUSSION TITLE
VBR in path_formats / queries / inline plugin

### DISCUSSION THREAD

**Discussion by JOJ0** [beetbox/beets#4624] [Q&A] [answered]
Is it possible to find out whether an MP3 file is VBR-encoded? I'd like to extend an inline plugin item_field to also return 1 if the MP3 file was encoded with VBR:
```
item_fields:
    ...
    ...
    quality: |
        if format == 'FLAC':
            return 1
        else:
            if bitrate > 200000:
                return 1
        return 0

```

I tried the `encoder` var already but that often does not contain anything.

I'm also up for any other ideas to find out whether the files are of a certain quality.

**Comment by JOJ0** [comment]
Using `mutagen-inspect`, I find this information in my files:

```
- MPEG 1 layer 3, 170297 bps (VBR, LAME 3.93.0+, --preset standard), 44100 Hz, 2 chn, 312.91 seconds (audio/mp3)
```

Is it possible to access the information in the first brackets (...) from beets?

**Comment by JOJ0** [comment]
I just found out that I can access a field called `bitrate_mode` that contains the `VBR` information:

```
$ beet info -i bitrate_mode united future good luck
/remote/data/music-library/DowntempoTriphopDubAcidjazz/United Future Organization - 1999 -  Bon Voyage/01 - United Future Organization - Bon Voyage - Good Luck Shore (1999).mp3
bitrate_mode: VBR
```

Unfortunately my extended inline plugin function does not work. Why?

```
    quality: |
        if format == 'FLAC':
            return 1
        elif bitrate > 255000:
            return 1
        elif 'VBR' in bitrate_mode:
            return 1
        return 0
````

I also tried to access this field with `beet ls`, but no luck:


```
$ ls -f '$bitrate_mode' united future good luck

$
```

It looks like the field is empty. If the variablename would be wrong the result would look like this:

```
$ beet ls -f '$bitratemode' united future good luck 
$ $bitratemode
```

Why? Any hints? Or bug? Or missing feature?

Cheers in advance!

**Comment by JOJ0** [comment, marked as answer]
Found a partly solution but not the root cause:

The bitrate_mode field often is empty in my collection (Total in lib about 2500):

```
$ beet ls bitrate_mode::'^$' format:mp3 | wc -l
1527
```

Trying to update from the files into the beets db:

```
beet update -F bitrate_mode bitrate_mode::'^$' format:mp3
```


results in: some files contain the information and beets lib is updated:

```
...
Rockers Hi‐Fi - Rockers to Rockers - 08 - Look for a Spark (1995/GB) [524 102-2]  f:MP3-119k a:02-12/MusicBrainz |Dub, Ska, Reggae, Downtempo
  bitrate_mode:  -> VBR
Asian Dub Foundation - Rafi’s Revenge - 01 - Naxalite (1998/GB) [FFRR / 556 006-2]  f:MP3-225k a:02-06/MusicBrainz |Dub, Ska, Reggae
  bitrate_mode:  -> VBR
Asian Dub Foundation - Rafi’s Revenge - 02 - Buzzin’ (1998/GB) [FFRR / 556 006-2]  f:MP3-251k a:02-06/MusicBrainz |Dub, Ska, Reggae
  bitrate_mode:  -> VBR
Asian Dub Foundation - Rafi’s Revenge - 03 - Black White (1998/GB) [FFRR / 556 006-2]  f:MP3-228k a:02-06/MusicBrainz |Dub, Ska, Reggae
  bitrate_mode:  -> VBR
...
```

Some files don't seem to contain the information and bitrate_mode stays empty.

It's unclear why those files that contain it, haven't been updated on beets import. Not sure if this is of interest to anyone but if it is: Could this be a bug? Or is there certain situation when it is expected behaviour that bitrate_mode is not written to the lib/database on import?




After beet update run was through only a few bitrate_mode fields could be filled (~50):
```
$ beet ls bitrate_mode::'^$' format:mp3 | wc -l             
1466
```

**Reply by RollingStar** [reply]
I've seen people use the heuristic of "is the bitrate 160, 192, 256, 128" etc. kbps and if not, assume VBR.

**Reply by JOJ0** [reply]
Thanks for this nice suggestion @RollingStar 

For now I came to the conclusion, well rather an assumption that CBR or VBR information might have been written by encoders already, and I suppose writing CBR is not always done, but VBR _is_ written, since is "something special" an encoder might want to mention. Could be bollocks but for now seems to work out :-)

FWIW, these are my current inline fields:
```
    quality: |
        # print(dir(bitrate_mode))
        # print(bitrate_mode)
        if format == 'FLAC':
            return 1
        elif bitrate >= 193000:
            return 1
        elif bitrate_mode == 'VBR':
            return 1
        return 0
    fmt_format: |
        f = 'f:'
        f += f'{format}'
        if format == 'MP3':
            f+= '-' + str(bitrate)[:3] + 'k'
        f += 'VBR' if bitrate_mode == 'VBR' else ''
        f += 'CBR' if bitrate_mode == 'CBR' else ''
        return f
```
